### PR TITLE
Initial commit to assignment 1

### PR DESCRIPTION
**Pull request by ickbumk**

### CODE REVIEW

File: polynomial_function.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <vector>
+#include "scalarfunction.hpp"
+
+// initialization of polynomial function class. The operator from the base scalarfunction class is overriden.
+// Constructs the polynomial function based on the coefficients provided by the user.
+
+class polynomial_function : public scalar_function{
+    public:
+    polynomial_function(const std::vector<double> & coefficients) : coefficients_(coefficients){}
```
Comment (gicamikiel) [inline polynomial_function.hpp:12]: Is there a reason why parameter coefficients is being passed by reference with the & symbol?

If the vector object in coefficients is modified outside of polynomial_function it might affect the output of polynomial_function.

File: integration_rule.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    public:
+    // constructor initialization of the integration nodes (sampling locations and weight pairs)
+    integration_rule(const std::vector<double>& nodes, const std::vector<double> & weights): nodes_(nodes), weights_(weights){
+        if (nodes_.size() != weights_.size())
+        throw std::invalid_argument("Nodes and weights different size");
+    }
+
+    const std::vector<double>&nodes() const { return nodes_;}
+
+    const std::vector<double>&weights() const {return weights_;}
```
Comment (gicamikiel) [inline integration_rule.hpp]: Returning the whole weights and nodes vector like this isn't very "defensive". 

If a reference to weights and nodes is returned like this, another developer would be able to add to nodes, but forget to add to weights, making the nodes and weights different size.

You want getters and setters that can properly check if the modifications being made to nodes and weights are valid and throw and error if they are not.

File: scalarfunction.hpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#ifndef SCALAR_FUNCTION_H
+#define SCALAR_FUNCTION_H
+
+
+#include <vector>
+#include <utility> // this enables the usage of std::pair for storing (sample location, weight)
```
Comment (gicamikiel) [inline scalarfunction.hpp]: Did you end up using std::pair anywhere?

File: polynomial_integrate.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    std::cout<< "Enter the type of integration rule (e.g. gle(gauss_legendre), glo(gauss-lobotto), cs(chebyshev)): ";
+    std::getline(std::cin, user_input);
+
+    std::cout<< "Enter the number of points: ";
+    std::cin >> c;
+
+    std::cout << "Starting the integration \n";
+
+    if(user_input == "gle"){
```
Comment (gicamikiel) [inline polynomial_integrate.cpp:67]: (minor style comment) In other files there is a space between the if and the (

File: unittest.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    double midpoint = (a+b)/2.0;
+    double result = 0.0;
+
+    const auto& nodes = rule.nodes();
+    const auto& weights = rule.weights();
+    for (size_t i =0; i<nodes.size(); ++i){
+        double x = scale * nodes[i] + midpoint;
+        result += weights[i] * func(x);
+    }
+    return scale * result;
```
Comment (gicamikiel) [inline unittest.cpp:26]: I think this functionality should be in a library instead of a source file for an executable. This process is required by anyone doing integration and if they need the integrator classes they will probably need this too.

See also comment on [polynomial_integrate.cpp](https://github.com/ickbumk/cpp/pull/2/files#diff-487f3dae3a15d655124c44e91dc470a79bd41e6a0d1ed1da42eb6338e6a11bb3R28)

File: polynomial_integrate.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    double midpoint = (a+b)/2.0;
+    double result = 0.0;
+
+    const auto& nodes = rule.nodes();
+    const auto& weights = rule.weights();
+    for (size_t i =0; i<nodes.size(); ++i){
+        double x = scale * nodes[i] + midpoint;
+        result += weights[i] * func(x);
+    }
+    return scale * result;
```
Comment (gicamikiel) [inline polynomial_integrate.cpp:28]: Repeated functionality from [unittest.cpp](https://github.com/ickbumk/cpp/pull/2/files#diff-0e1974fb145d23c3ceec36d8f16e5d529d505fc6eeec4211a68b53f0bd51e846). If this code belongs to a library alongside the integration rule and scalar function classes duplication would not be needed.

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.28.3)
+project(assignment1 CXX)
+
+add_executable(polynomial_integrate polynomial_integrate.cpp)
+add_executable(unittest unittest.cpp)
+
+
+target_include_directories(polynomial_integrate PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}$)
```
Comment (gicamikiel) [inline CMakeLists.txt:8]: According to the assignment we need to "_create a library for your integration routines_
and two executables" (assignment 1 item 1e).

CMakeLists.txt is making only the two executables and naively including the integration and scalar function code in their compilations. I do not think any libraries are being made here (using add_library).

This is not really a good way to do things because other developers cannot use the integration and scalar function code without compiling it alongside their own code. Having a library being compiled separately then linked is much more convenient, which is why I think the assignment specifies to make a library.

File: unittest.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    b = {1.0};
+    c = {4};
+
+    std::cout<< "The polynomial coefficients chosen for this is 2,3,4,5: \n";
+    std::vector<double> coefficients ={4,3,2,1};
+    
+    std::cout << "Starting the integration \n";
+
+    try {
+        test_glo(coefficients, a, b, c, get_gauss_lobotto(c));
```
Comment (gicamikiel) [inline unittest.cpp:81]: I think there should be unit tests for the scalar function classes.

File: unittest.cpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+int main(){
+    std::cout<< "Unittest will be carried with integration limits 0 and 1: \n";
+    double a, b, c;
+
+    a = {0.0};
+    b = {1.0};
+    c = {4};
+
+    std::cout<< "The polynomial coefficients chosen for this is 2,3,4,5: \n";
+    std::vector<double> coefficients ={4,3,2,1};
```
Comment (gicamikiel) [inline unittest.cpp:76]: Why/how were these coefficients picked for testing?

File: unittest.cpp [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    
+    std::cout << "Starting the integration \n";
+
+    try {
+        test_glo(coefficients, a, b, c, get_gauss_lobotto(c));
+        test_gle(coefficients, a, b, c, get_gauss_legendre(c));
+        test_cs(coefficients, a, b, c, get_chebyshev(c));
+        return 0;
```
Comment (gicamikiel) [inline unittest.cpp:84]: I think there should be a message when the test passes. Currently it seems like something has silently failed.

File: unittest.cpp [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    } catch (...){
+        return 1;
```
Comment (gicamikiel) [inline unittest.cpp:86]: This works but in class I believed we discussed that the return type should be the number of tests that failed. I think this is so any automated testing process running this executable can get more information.

**Review by gicamikiel — APPROVED** [review]
The code is neat and easy to understand. I think there are some issues about classes and how CMake is configured.

**Does the style match?**

Apart from some very minor spacing, yes.

**Is there documentation of the new code? Does it address what it does and how to use it?**

There are only a couple of lines of comments for documentation, and these are enough.

**Are there any obvious bugs, or performance issues?**

**Is the code easy to read and understand? Is it well organized? Are function names clear?**

Yes for all three questions. There have been no cryptic symbols or names.

**Can raw loops be replaced with standard library algorithms?**

Could have replaced one loop for computing the exponential with a std::pow from the math library, but this is very minor

**General questions or concerns.**

The two most serious concerns are about the CMake config and class get and set functions. See comments.

**Does the code compile?**

The code compiles

**Are there any compiler warnings?**

No warnings

**Are there any runtime errors?**

No runtime errors, erronous input seem to be handled well.

**Does the code do what it is supposed to do?**

Yes, I entered some polynomials where the quadratures are known to compute exact and it seems to work.

**Are there test cases for all new functionality?**

There are only unit tests for the integration classes

**Are there test cases that you think should be added?**

Unit tests for scalar function classes (polynomial function)

I am not sure if there is any specific reasons for the selection of integration classes unit tests, might be good to document with a few comments.

**Do all of the tests pass?**

Yes. It might be better if the unit test executable was more verbose about tests passing. No success message makes it seem like something has gone wrong.

**Are there any performance regressions?**

We have not gone over measuring performance yet so N/A

**Does the code scale to the expected problem size?**

The assignment does not specify anything demanding to integrate so N/A

Most of my experience with writing classes comes from a lot of abusing OOP in C# as a hobby, so feel free to correct me if anything is not applicable to C++.

File: gauss_lobotto.hpp [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+            return integration_rule(nodes,weights);
+        }
+        case 4: {
+            double invsqrt5 =  1.0/std::sqrt(5.0);
+            std::vector<double> nodes = {-1.0, -invsqrt5, invsqrt5, 1.0};
+            std::vector<double> weights = {1.0/6.0,5.0/6.0,5.0/6.0, 1.0/6.0};
+            return integration_rule(nodes,weights);
+        }
+        case 5: {
+            double sqrt37 = std::sqrt(3.0/7.0);
```
Comment (scottblender) [inline gauss_lobotto.hpp:26]: This is a confusing name. I might name it something else that is more descriptive. Same with the other variable names.

File: integration_rule.hpp [hunk 13]
```diff
@@ -0,0 +1,28 @@
+#ifndef INTEGRATIONRULE_HPP
+#define INTEGRATIONRULE_HPP
+
+#include <vector>
+#include <stdexcept>
+
+// Class initialization for the integration assignment
+// Integration rule can be specified with different methods (gauss legendre, gauss lobotto, chebyshev)
+
+class integration_rule {
```
Comment (scottblender) [inline integration_rule.hpp:11]: I might recommend using templates when initializing the std::vector. When making a newly derived class, what if someone tries initializing weights as an integer or nodes as a double?

File: polynomial_integrate.cpp (hunk 6 above)
Comment (scottblender) [reply polynomial_integrate.cpp:28]: ^^ I would make this its own file and write a header for it. Make it a library and use `target_link_libraries` as we did with math library.cpp in the in-class example.

File: polynomial_integrate.cpp [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    double a, b, c;
+    std::string user_input;
+
+
+    std::cin >> a >> b;
+    std::cin.ignore(); //discard the rest of the line
+
+    std::cout<< "Enter polynomial coefficients (c0 c1 c2 ...) separed by space: ";
+    std::vector<double> coefficients = parseDoubles();
+    std::unique_ptr<scalar_function> polyfunc = std::make_unique<polynomial_function>(coefficients);
```
Comment (scottblender) [inline polynomial_integrate.cpp:57]: Is this the best way to initialize a polynomial?

File: polynomial_integrate.cpp [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        std::cout<< "user selected Chebyshev integration method\n";
+
+        integration_rule rule = get_chebyshev(c);
+        double approx = integrate(*polyfunc, a, b, rule);
+        std::cout << "Chebyshev integration result is approximately: "<< approx << "\n";
+
+    }else {
+        std::cout<<"Incorrect integration method specified \n";
+    }
+
```
Comment (scottblender) [inline polynomial_integrate.cpp:93]: return 0?

File: unittest.cpp (hunk 11 above)
Comment (scottblender) [reply unittest.cpp:86]: Yeah, I would just sum the number of tests that passed or failed.

File: unittest.cpp (hunk 9 above)
Comment (scottblender) [reply unittest.cpp:76]: There needs to be more robust testing. I would also ensure that the polynomial entered can be integrated using the specified method. Gauss quadrature is only valid for polynomials of up to 2*n-1.

**Review by scottblender — COMMENTED** [review]
- Does the style match the rest of the codebase? (naming conventions, indentation, etc.) Tools like clang-format can help.
    -The style is mostly consistent. Variable names can be improved to make them clearer to someone compiling the code, specifically for the coefficients in the derived classes.

- Is there documentation of the new code? Does it address what it does and how to use it?
    - Yes, the documentation is clear and sufficient.

- Are there any obvious bugs or performance issues?
    - No, there are no obvious bugs. I would reduce redundancy and make a library of repeated functions used though to leverage CMake better.

- Is the code easy to read and understand? Is it well-organized? Are function names clear?
    - Yes, the code is easy to read and well-organized.
  
- Can raw loops be replaced with standard library algorithms?
    - To my knowledge, no. The `std::for_each` might be worthwhile to check out, though. 

- Does the code compile?
    - Yes, it compiles without errors.

- Are there any compiler warnings?
    - No, there are no warnings.

- Are there any runtime errors?
    - No, there are no runtime errors.

- Does the code do what it is supposed to do?
    - The code does not exactly do what it is supposed to. The polynomial_integrate works as expected, but unit_tests does not output what it should. The goal of unit_tests is to be able to print the sum of how many tests failed and the result of each test. What I did was test polynomials where I knew the analytic solution to so I could check relative error.

- Are there test cases for all new functionality?
    - There are checks for the number of integration points and user inputs. I might recommend adding templates and also checking to make sure that the polynomial entered is valid/can be integrated by the specified quadrature method.

- Are there test cases that you think should be added? i.e., corner cases that are not covered.
    - I would check a few more polynomials.

- Do all of the tests pass?
    - I'm not sure, since the unit_tests does not ouput results for each test.

- Are there any performance regressions?
    - No, performance has not been impacted.

- Does the code scale to the expected problem size?
    - For this, the code does not need to be scalable, but it looks like it would work fine.

File: scalarfunction.hpp (hunk 3 above)
Comment (ickbumk) [reply scalarfunction.hpp]: I initially used for this another version. Thank you for pointing out

File: polynomial_function.hpp (hunk 1 above)
Comment (ickbumk) [reply polynomial_function.hpp:12]: I think passing coefficients by reference instead of by value is more efficient. 
Instead of value (std::vector<double> coefficients) which will create a copy of the vector, I used const reference (const std::vector<double> & coefficients) to avoid extra copy.
Const ensures it cannot be modified outside.

File: integration_rule.hpp (hunk 2 above)
Comment (ickbumk) [reply integration_rule.hpp]: Thank you for the recommendation. I made the changes accordingly.

**Comment by ickbumk** [comment]
I have made some changes to the code. I'm left with changing unittest, which is taking longer than I think.
